### PR TITLE
Removed service worker

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This removes all of our service worker functionality with a self-destructing service worker which appears to be the best approach for unregistering a service worker. See https://github.com/NekR/self-destroying-sw#how-to-use.

## Related Tickets & Documents

#12899
#12631
#12455
#12000
#11606 
#11533
#11488
#11270
#10293
#10173
#9657 
#9642 
#9509
#9254
#9228
#8729
#7339
#7081
#6497

## QA Instructions, Screenshots, Recordings

**Test this in Chrome, Firefox, and Safari. Note in Safari, there may be no service worker registered due to the work done with the service-companion.js file. In that case, ensure that no errors are thrown in the console of the web dev tools in Safari.**

1. Pull down latest of the main branch and deploy the site locally. Navigate to the home page and validate that the service worker is running by opening the developer tools and navigating to the developer tools.

**Chrome**

image

**Firefox**

image

**Safari**

image

2. Pull down the branch for this PR and deploy it locally. Refresh the browser window that is currently open and validate that the service worker gets unregistered.

**Chrome**

image

**Firefox**

image

**Safari**

image

### UI accessibility concerns?

N/A this is front-end infrastructure

## Added tests?

- [ ] Yes
- [x] No, and this is why: I've removed tests related to the service worker, so existing tests should continue to pass.
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [x] @mstruve will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

Samuel L. Jackson in Jurassic Park saying "Hold on to your butts"

### CODE REVIEW

File: app/assets/javascripts/base.js.erb
Comment (nickytonline) [inline app/assets/javascripts/base.js.erb:8]: This was previously used for Mac/iOS, but is no longer required since the self-destructing service worker does what this file was doing.

File: app/controllers/application_controller.rb
Comment (nickytonline) [inline app/controllers/application_controller.rb:49]: These shell partials are now part of application.html.erb

File: app/lib/url.rb
Comment (nickytonline) [inline app/lib/url.rb:86]: It says this is for the service worker, but do we still want to keep the check for `referrer.blank?` ?

File: app/views/layouts/application.html.erb
Comment (nickytonline) [inline app/views/layouts/application.html.erb:1]: top and bottom shell pieces are now part of this file.

File: app/views/service_worker/index.js.erb
Comment (nickytonline) [inline app/views/service_worker/index.js.erb:5]: The most reliable way to unregister a service worker is to create a new one that unregisters itself according to https://github.com/NekR/self-destroying-sw

**Comment by nickytonline** [comment]
I've got some rspec test failures. I can look at those later tonight unless someone wants to jump on that.

File: app/views/service_worker/index.js.erb
Comment (fdoxyz) [inline app/views/service_worker/index.js.erb:79]: Does this mean we will no longer have the fallback offline (doodle) page or will it continue to work but just not triggered from here?

**Review by fdoxyz — COMMENTED** [review]
Haven't had a chance to test it yet but left an initial question inline

File: app/lib/url.rb
Comment (nickytonline) [reply app/lib/url.rb:86]: Thanks @citizen428. I've made the changes.

File: app/lib/url.rb
Comment (citizen428) [inline app/lib/url.rb:74]: Sorry if my original comment was unclear, but we no longer need this due to the other 2 changes. They were the only two call sites of this method and I think now that we no longer have the service worker guard clause this method doesn't add much.

File: app/views/service_worker/index.js.erb
Comment (nickytonline) [reply app/views/service_worker/index.js.erb:79]: @fdoxyz, the offline page was a feature of the service worker, so the only way you'd be able to access the page now is by going directly to e.g. http://localhost:3000/offline.html

Comment (Zhao-Andy) [reply app/views/service_worker/index.js.erb:79]: Should we also remove that file while we're here? Could always recover it later if we want to bring it back.

**Comment by nickytonline** [comment]
Just need to sort out these test failures:

```bash
Failures:

  1) Pages GET /report-abuse when provided the referer prefills with the provided url

     Failure/Error: reported_url = params[:reported_url] || params[:url] || referer.presence

     NameError:

       undefined local variable or method `referer' for #<PagesController:0x00007f76225f7810>

     # ./app/controllers/pages_controller.rb:86:in `report_abuse'

     # ./app/middlewares/set_time_zone.rb:10:in `call'

     # ./app/middlewares/set_cookie_domain.rb:13:in `call'

     # ./spec/requests/pages_spec.rb:246:in `block (4 levels) in <top (required)>'

     # ./spec/[secure]_helper.rb:151:in `block (3 levels) in <top (required)>'

     # ./spec/[secure]_helper.rb:151:in `block (2 levels) in <top (required)>'

  2) feedback_messages POST /feedback_messages with valid params and recaptcha not configured does not show the recaptcha tag

     Failure/Error: reported_url = params[:reported_url] || params[:url] || referer.presence

     NameError:

       undefined local variable or method `referer' for #<PagesController:0x00007f7620c24b18>

     # ./app/controllers/pages_controller.rb:86:in `report_abuse'

     # ./app/middlewares/set_time_zone.rb:10:in `call'

     # ./app/middlewares/set_cookie_domain.rb:13:in `call'

     # ./spec/requests/feedback_messages_spec.rb:89:in `block (4 levels) in <top (required)>'

     # ./spec/[secure]_helper.rb:151:in `block (3 levels) in <top (required)>'

     # ./spec/[secure]_helper.rb:151:in `block (2 levels) in <top (required)>'
```

File: app/controllers/service_worker_controller.rb
Comment (joshpuetz) [inline app/controllers/service_worker_controller.rb:8]: Do we need this controller at all anymore, or is it used for 'deleting' existing service workers?

**Review by joshpuetz — APPROVED** [review]
Tested locally in chrome and successfully removed an already installed service worker. 👍🏻

File: app/controllers/service_worker_controller.rb
Comment (nickytonline) [reply app/controllers/service_worker_controller.rb:8]: We need it still for the self destructing service worker.

File: app/views/service_worker/index.js.erb
Comment (nickytonline) [reply app/views/service_worker/index.js.erb:79]: Sounds good @Zhao-Andy. Will do.

Comment (nickytonline) [reply app/views/service_worker/index.js.erb:79]: It's been removed. cc: @fdoxyz.

File: app/controllers/pages_controller.rb
Comment (fdoxyz) [inline app/controllers/pages_controller.rb:86]: The referer needs to be accessed via the request object (in the past the plain `referer` was the variable that no longer exists). I think this one is the reason a spec is failing.

Comment (nickytonline) [reply app/controllers/pages_controller.rb:86]: I went a little overboard with the deletion. 🙃

File: app/controllers/pages_controller.rb
Comment (fdoxyz) [inline app/controllers/pages_controller.rb]: Suggestion:

```ruby
reported_url = params[:reported_url] || params[:url] || request.referer.presence
```

File: app/helpers/application_helper.rb
Comment (fdoxyz) [inline app/helpers/application_helper.rb]: Suggestion:

```ruby
request.referer.presence
```

File: app/lib/url.rb
Comment (fdoxyz) [inline app/lib/url.rb:85]: If we revert the suggestion we do need this method because it's called from `application_helper.rb` and `pages_controller.rb `

**Review by fdoxyz — COMMENTED** [review]
Sorry @nickytonline I feel like I could've been a bit more specific too!

@citizen428's suggestion was good IMO, but we just needed to prepend `request.referer.presence` instead of `referer.presence` to the two places where they were modifed (right now the `URL. sanitized_referer` method doesn't exist). I left the suggestions inline.

Either that or adding the `URL. sanitized_referer` method again so specs pass. A cleanup on that can be done later on if we want to go with that suggestion after all.

**Review by nickytonline — COMMENTED** [review]
Hmm, GitHub is not showing two last commits even though I pushed succesfully. 🤔

**Comment by jgaskins** [comment]
@nickytonline GitHub's having a pretty severe outage. They're currently recovering from it, so it'll probably take a while for them to work through their backlog.

**Review by fdoxyz — APPROVED** [review]
I tested it out a bit, including authentication on multiple providers in Chrome & Safari. LGTM 👍 

The last commit brought something to attention that I hadn't considered, all I have is a **non-blocking question**:

Is there a way we can keep the `manifest.json` even with the self unregistering service worker? So users can still install the PWA on the platforms that do support it.

**Comment by nickytonline** [comment]
> I tested it out a bit, including authentication on multiple providers in Chrome & Safari. LGTM 👍
> 
> The last commit brought something to attention that I hadn't considered, all I have is a **non-blocking question**:
> 
> Is there a way we can keep the `manifest.json` even with the self unregistering service worker? So users can still install the PWA on the platforms that do support it.

@fdoxyz, I'm not sure. I'm not too familiar with PWAs. I can add it back and see if it works even with a self-destructing service worker.

**Comment by nickytonline** [comment]
@fdoxyz, looks like a SW is required. I ran Lighthouse and the screenshot is complaining about the missing manifest, but it does indicate a SW is required for a PWA.

image

**Comment by fdoxyz** [comment]
> @fdoxyz, I'm not sure. I'm not too familiar with PWAs. I can add it back and see if it works even with a self-destructing service worker. cc: @maestromac since you mentioned this will remove the PWA.

Absolutely, my question comes out of not being too familiar with PWAs as well 😄 

> @fdoxyz, looks like a SW is required. I ran Lighthouse and the screenshot is complaining about the missing manifest, but it does indicate a SW is required for a PWA.

Makes sense. I'm still totally okay with going forward with this so we can look into a well thought out plan of action to get PWAs working again (including @forem/product  in the conversation too since it could take the RFC route).

Just wanted to share why I would personally rather wait until next week to merge this: Since the PR removes PWA support and users could reach out over the weekend maybe worth waiting for: ` I will share this change in a Changelog or in a forem.dev post`. Also, the SW functionality has "settled" a bit from the patches added over the recent incidents and I don't want to "poke the bear" on Friday afternoon 😅

**Comment by nickytonline** [comment]
> Makes sense. I'm still totally okay with going forward with this, and maybe we can look into a well thought out plan of action to get PWAs working again (including @forem/product in the conversation too since it could take the RFC route).

I definitely think we can bring it back, but an RFC makes sense. Also, it'll be good to let the dust settle a bit. 😅 

> Just wanted to share why I would personally rather wait until next week to merge this: Since the PR removes PWA support and users could reach out over the weekend maybe worth waiting for: ` I will share this change in a Changelog or in a forem.dev post`. Also, the SW functionality has "settled" a bit from the patches added over the recent incidents and I don't want to "poke the bear" on Friday afternoon 😅

I'm totally OK with waiting until next week as well. cc: @mstruve

**Review by maestromac — APPROVED** [review]
Tested locally, yup it's removed on my browser :+1:  . Nice job @nickytonline !

the PR itself is LGTM but maybe need to run test it on staging fiirst.

**Comment by liflovs** [comment]
@nickytonline does this issue help with seo problem in https://github.com/forem/forem/issues/9509?

**Comment by nickytonline** [comment]
> @nickytonline does this commit help with seo problem in https://github.com/forem/forem/issues/9509?

@liflovs , it should but I think there may be some additional dead code to remove base on @benhalpern's comment https://github.com/forem/forem/issues/9509#issuecomment-668888537. Going to out this back into draft mode to validate this next week.

**Comment by citizen428** [comment]
Tested with Chromium Edge on Windows, works as expected here too 👍

**Review by citizen428 — APPROVED** [review]

**Comment by Zhao-Andy** [comment]
Just a heads up that I restarted the failed build Travis but maybe I didn't need to 😅

**Review by mstruve — APPROVED** [review]
🚀

**Pull request merged**
